### ISSUE TITLE
12 Main Dropout Methods: Mathematical and Visual Explanation

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6791]
12 Main Dropout Methods&nbsp;: Mathematical and Visual Explanation<br>
<br>
https://ift.tt/2zZ1Wov<br>
<br>
Axel Thevenot<br>
<br>
<div>
<div>
<div>
<p>We can, therefore, see that the greater the <strong>weight</strong>, the greater the <strong>probability</strong> that the neuron will be omitted. This powerfully limits the <strong>high predictive capacity</strong> that some neurons may have.</p>
<h1>Gaussian Dropout</h1>
<p>The list of dropout methods applied to neural networks continues to grow. So before moving on to something else than DNNs I would like to talk about a category of Dropout method that is certainly the most <strong>fascinating</strong>.</p>
<p>To take just a few examples, <strong>Fast Dropout</strong> [4], <strong>Variational Dropout</strong> [5] or <strong>Concrete Dropout</strong> [6] are methods interpreting dropout from a <strong>Bayesian perspective</strong>. Concretely, instead of having a Bernoulli mask, we have a mask whose elements are random <strong>variables following a Gaussian distribution</strong> (<strong>Normal</strong> distribution). I won&rsquo;t go into the demonstration of the <strong>law of Large Numbers</strong> here, that is not the point. So let&rsquo;s try to understand this <strong>intuitively</strong>.</p>
<figure>
<figcaption>Dropout with p=0.5</figcaption>
</figure>
<p>Papers [4], [5] and [6] show we can <strong>simulate</strong> a Bernoulli mask for our dropouts with a <strong>normal law</strong>. But what difference does it make. <strong>Everything and nothing</strong> at the same time. It does not change anything concerning the relevance of these methods against <strong>overfitting</strong> due to <strong>co-adaptation</strong> and/or the <strong>predictive capacity</strong> of our neurons. But it changes everything in terms of the <strong>execution time required for the training</strong> phase compared to the methods presented before.</p>
<p>Logically, by omitting at each iteration neurons with a dropout, <strong>those omitted on an iteration are not updated</strong> during the backpropagation. They do not exist. So the training phase is <strong>slowed down</strong>. On the other hand, by using a Gaussian Dropout method, <strong>all the neurons are exposed at each iteration</strong> and for each training sample. This avoids the slowdown.</p>
<p>Mathematically, there is a multiplication with a Gaussian mask ( for example centered in 1 with Bernoulli&rsquo;s law standard deviation <code>p(1-p)</code>). This <strong>simulates</strong> the dropout by randomly <strong>weighting their predictive capacity</strong> by keeping all neurons <strong>active</strong> at each iteration. Another practical advantage of this method centered in 1: during the testing phase there is no modification to be made compared to a model without dropout.</p>
<h1>Pooling Dropout</h1>
<p>The &ldquo;difficult&rdquo; comprehension part of this article is over. Remaining the more <strong>intuitive</strong> part giving to us <strong>better performances</strong>.</p>
<p>The issue with <strong>images</strong> or <strong>feature maps</strong> is that pixels are very <strong>dependent on their neighbors</strong>. To put it simply, on a cat picture, if you take a pixel that corresponds to its coat then all the neighboring pixels will correspond to the same coat. There is little or <strong>no difference</strong>.</p>
<p>So we understand the <strong>limits</strong> of the Standard Dropout method. We could even say it is <strong>inefficient</strong> and the only change it brings is <strong>additional computation time</strong>. If we randomly omit pixels on an image then almost no information is removed. The <strong>omitted pixels are nearly the same as their surroundings</strong>. It means <strong>poor</strong> performance to prevent overfitting.</p>
<p>Why not to take advantage of the layers which are <strong>proper</strong> and often used in the <strong>CNNs</strong>. For example the <strong>Max Pooling Layer</strong>. For those who don&rsquo;t know: the Max Pooling Layer is a filter passed on a picture or (feature map) selecting the maximum activation of the overlapping region.</p>
<p><strong>Max-Pooling Dropout</strong> [7] is a dropout method applied to CNNs proposed by H. Wu and X. Gu. It applies Bernoulli&rsquo;s mask directly to the <strong>Max Pooling Layer</strong> kernel <strong>before</strong> performing the pooling operation. Intuitively, this allows minimizing the pooling of high activators. It is a very good point to <strong>limit the heavy predictive capacity</strong> of some neurons. During the test phase, you can then weight as for the previous methods by the probability of presence.</p>
<p>The Max Pooling Layer has been taken as an example, but the same could be done with <strong>other Pooling Layers</strong>. For example, with the <strong>Average Pooling Layer</strong>, we could apply a dropout in the same way during the training phase. Then in the test phase, there would be no change since it is already a weighted average.</p>
<h1><strong>Spatial Dropout</strong></h1>
<p>For the CNNs, we can take advantage of the Pooling Layers. But we can also go smarter by following the <strong>Spatial Dropout</strong> [8] method proposed by J. Tompson et al. They propose to <strong>overcome the problem</strong> with classical dropout methods because the <strong>adjacent pixels</strong> are highly <strong>correlated</strong>.</p>
<p>Instead of randomly applying a dropout on the pixels we can think about <strong>applying a dropout per feature map</strong>. If we take the example of our cat, then this is like removing the red from the image and forcing it to generalize on the blue and green of the image. Then randomly other feature maps are dropped on the next iterations.</p>
<p>I did not know how to write properly in mathematics to make it intelligible. But if you understood the previous methods, you won&rsquo;t have any trouble with it. In the <strong>training</strong> phase, a <strong>Bernoulli mask</strong> is applied <strong>per feature map</strong> with a probability of omission <code>p</code>. Then during the <strong>testing</strong> phase, there is no dropout but a <strong>weighting</strong> by the probability of presence <code>1-p</code>.</p>
<h1>Cutout</h1>
<p>Let&rsquo;s go <strong>deeper</strong> into our approach to overcome the fact that adjacent pixels are highly correlated. Instead of applying Bernoulli masks per feature map, they can be <strong>applied in areas</strong>. This is the <strong>Cutout</strong> method [9] proposed by T. DeVries and G. W. Taylor.</p>
<p>By taking one last time the example of our cat image: this method makes it possible to generalize and thus limit overfitting by <strong>hiding areas of the image</strong>. We end up with images where the cat&rsquo;s head is dropping. This forces the CNN to recognize the less obvious attributes describing a cat.</p>
<p>Again in this section no math. This method depends a lot on our imagination: square areas, rectangles, circles, on all feature maps, on one at a time or possibly on several&hellip; It&rsquo;s <strong>up to you</strong>. &#128515;</p>
<h1>Max-Drop</h1>
<p>Finally, to conclude this section on the CNNs, I must point out that obviously several <strong>methods can be combined</strong>. This is what makes us strong when we know the different methods: we can take <strong>advantage of their benefits at the same time</strong>. This is what S. Park and N. Kwak propose with their <strong>Max-Drop</strong> method [10].</p>
<p>This approach is in a way a <strong>mixture of Pooling Dropout and Gaussian Dropout</strong>. The dropout is performed on the <strong>Max Pooling Layer</strong> but with a <strong>Bayesian approach</strong>.</p>
<p>In their paper, they show that this method gives results as <strong>efficient</strong> as with a Spatial Dropout. In addition to the fact that at each iteration, all the neurons remain activated which <strong>limits the slowdown</strong> during the training phase.<br>
These results were obtained with &micro; = 0.02 and &sigma;&sup2; = 0.05 .</p>
<h1>RNNDrop</h1>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Medium<br>
<br>
June 12, 2020 at 02:01PM
